src/parser/queries/QueryBuilder.ts

```diff
@@ -36,25 +42,57 @@ export class QueryBuilder {
 
         if (this.#language !== Language.Unknown) {
             if (this.#statements.length > 0) {
-                queryString = this.getStatementsQuery();
+                queryString = this.#getStatementsQuery();
             }
 
             dlog("------------- Start Query: --------------");
             dlog(queryString);
             dlog("-----------------------------------------");
 
-            return new Query(this.#treeSitterLanguage, queryString);
+            return this.#retrieveTreeSitterQuery(queryString);
         }
     }
 
+    /**
+     * Tries to retrieve the tree-sitter query from the cache. If there is no matching query,
+     * calls the constructor of tree-sitter.Parser.Query, which is a very performance-heavy operation.
+     */
+    #retrieveTreeSitterQuery(queryString: string): Query {
+        let cachedQueriesForLanguage = treeSitterQueryCache.get(this.#language);
+
+        if (cachedQueriesForLanguage !== undefined) {
+            const cachedQuery = cachedQueriesForLanguage.get(queryString);
+            if (
+                cachedQuery !== undefined &&
+                cachedQuery.queryString === queryString &&
+                cachedQuery.language === this.#language
+            ) {
+                return cachedQuery.query;
+            }
+        }
+        const newQuery = new Query(languageToGrammar.get(this.#language), queryString);
+
+        if (cachedQueriesForLanguage === undefined) {
+            cachedQueriesForLanguage = new Map<string, CachedQuery>();
+            treeSitterQueryCache.set(this.#language, cachedQueriesForLanguage);
+        }
+        treeSitterQueryCache.get(this.#language)?.set(queryString, {
+            language: this.#language,
+            queryString: queryString,
+            query: newQuery,
+        });
```

Comment:
I think you could change the order of the two blocks.
After adding a new query, you can be sure that there will be one for the current lookup. Pseudo Code:

```suggestion
        let cachedQueriesForLanguage = treeSitterQueryCache.get(this.#language);
        
        if (cachedQueriesForLanguage === undefined) {
            cachedQueriesForLanguage = new Map<string, CachedQuery>();
            treeSitterQueryCache.set(this.#language, cachedQueriesForLanguage);
        }
        
        const newQuery = new Query(languageToGrammar.get(this.#language), queryString);
        if (newQuery not in treeSitterQueryCache.get(this#language)) {
          treeSitterQueryCache.get(this.#language)?.set(queryString, {
              language: this.#language,
              queryString: queryString,
              query: newQuery,
          });
        }
        
      
            return cachedQueriesForLanguage.get(queryString);

```

Reply:
This change suggestion would call the constructor of Query every single time, which is the thing this change tries to avoid.

Reply:
But I could, instead of returning early, go into an `else`-block for creating and adding a new query and retrieve the query from the cache again at the end of the function. That would be an unessesary additional time of retrieving an element from the map, however, which I am hesitant to do on such a performance-critical part of the code.

Reply:
I have refactored that while removing the CachedQuery interface as suggested below.